### PR TITLE
Fixing how the LocalAcccounts tests are skipped in Pester

### PR DESCRIPTION
**Pull request by mirichmo**
I updated the filter to conform to Pester guidance.

### CODE REVIEW

**Comment by vors** [comment]
@mirichmo can you provide a link to this guidelines?
I find this way of skipping them questionable and potentially leading to problems. Would like to discuss.

**Comment by lzybkr** [comment]
The guidance is in Jim's PR: https://github.com/JamesWTruher/PowerShell-1/blob/f8db6811a267e8cb7b58ee0b37181716d4317316/docs/testing-guidelines/WritingPesterTests.md

**Comment by vors** [comment]
The current approach is definitely better then what it used to be. The problem that I see is that it's too verbose. There are a lot of Describe blocks with few It blocks inside and it creates an overwhelmingly verbose massaging to achieve the goal. I think it's appropriate to use a modification of this technique here: it could be done once per file, at the beginning, and then the rest of the file can be wrapped in try-finally, where previous parameters are restored in finally. It's less idiomatic, but saves a lot of copy-paste.

**Comment by mirichmo** [comment]
The motivation for using this technique is that the tests are tracked consistently across all platforms. The original method caused these tests to disappear when run on non-Windows platforms. Now they can at least be tracked as skipped.

**Comment by vors** [comment]
Exactly, that's why I said it's better then it used to be. I'm just proposing to use a less verbose variation of the skipping, do it as a top-level thing on the file. @JamesWTruher can you comment?

**Comment by mirichmo** [comment]
I uncovered one wrinkle to this test skipping technique when fixing the tests. It becomes less clean when tests need to do actual work in the BeforeAll, AfterAll, BeforeEach, and AfterEach. This is what I settled on:
``` powershell
    BeforeAll {
        #skip all tests on non-windows platform
        $originalDefaultParameterValues = $PSDefaultParameterValues.Clone()
        $PSDefaultParameterValues["it:skip"] = ($IsWindows -eq $false)

        if ($PSDefaultParameterValues["it:skip"] -eq $false) {
            <work>
        }
    }
    AfterAll {
        if ($PSDefaultParameterValues["it:skip"] -eq $false) {
            <work>
        }
        $global:PSDefaultParameterValues = $originalDefaultParameterValues
    }
    BeforeEach {
        if ($PSDefaultParameterValues["it:skip"] -eq $false) {
            <work>
        }
    }

    AfterEach {
        if ($PSDefaultParameterValues["it:skip"] -eq $false) {
            <work>
        }
    }
```

**Comment by lzybkr** [comment]
We could use a convention like:
```powershell
BeforeAll {
        #skip all tests on non-windows platform
        $originalDefaultParameterValues = $PSDefaultParameterValues.Clone()
        $skipIt = ($IsWindows -eq $false)
        $PSDefaultParameterValues["it:skip"] = $skipIt

        if ($skipIt) {
            # work
        }
    }
    AfterAll {
        if ($skipIt) {
            # work
        }
        $global:PSDefaultParameterValues = $originalDefaultParameterValues
    }
    BeforeEach {
        if ($skipIt) {
            # work
        }
    }

    AfterEach {
        if ($skipIt) {
            # work
        }
    }
```

I'm not really a fan of `PSDefaultParameterValues` though, so if enough people prefer to explicitly specify `-Skip:$skipIt`, I think I'd +1 that proposal.

**Comment by vors** [comment]
I prefer explicit, because PSDefaultParameters contains global state and if something goes wrong it can mess-up the rest of execution royally. But for the bulk skip, explicit skip is too verbose.

**Comment by lzybkr** [comment]
Maybe the right solution is to work with Pester to provide `-Skip` on `Describe` in such a way that reports everything we want. In the shorter term, we can live with this solution.

**Comment by mirichmo** [comment]
@vors or @lzybkr it is ready for final review. Please take a look

**Comment by vors** [comment]
We talked offline about the approach with @JamesWTruher and he agreed that it's too verbose. Please implement the top level try-finally.

**Comment by mirichmo** [comment]
A top-level try-finally doesn't really shrink it all that much. This is what you end up with:

``` powershell
try {
  #skip all tests on non-windows platform
  $originalDefaultParameterValues = $PSDefaultParameterValues.Clone()
  $skipIt = ($IsWindows -eq $false)
  $PSDefaultParameterValues["it:skip"] = $skipIt

  Describe "Example for skipping a whole file" {
    BeforeAll {
        if ($skipIt) {
            # work
        }
    }
    AfterAll {
        if ($skipIt) {
            # work
        }
    }
    BeforeEach {
        if ($skipIt) {
            # work
        }
    }

    AfterEach {
        if ($skipIt) {
            # work
        }
    }
  }
}
finally {
  $global:PSDefaultParameterValues = $originalDefaultParameterValues
}
```
@JamesWTruher Please confirm this is what you want

**Comment by mirichmo** [comment]
@vors I made the try-catch change. It's ready for review now

File: test/powershell/Modules/Microsoft.PowerShell.LocalAccounts/Pester.Command.Cmdlets.LocalAccounts.LocalGroup.Tests.ps1 [hunk 1]
```diff
@@ -28,653 +22,689 @@ function VerifyFailingTest
...
+            }
         }
     }
-}
 
-Describe "Validate simple Get-LocalGroup" -Tags "CI", "Feature" {
-    BeforeAll {
-        New-LocalGroup -Name TestGroupGet1 -Description "Test Group Get 1 Description" | Out-Null
-    }
+    Describe "Validate simple Get-LocalGroup" -Tags "CI", "Feature" {
```
Comment (vors) [inline test/powershell/Modules/Microsoft.PowerShell.LocalAccounts/Pester.Command.Cmdlets.LocalAccounts.LocalGroup.Tests.ps1]: In #2366 you mentioned that you addressed it.....

File: test/powershell/Modules/Microsoft.PowerShell.LocalAccounts/Pester.Command.Cmdlets.LocalAccounts.LocalGroup.Tests.ps1 [hunk 2]
```diff
@@ -28,653 +22,689 @@ function VerifyFailingTest
...
+            $result -contains "Set-LocalGroup" | Should Be $true
+            $result -contains "Get-LocalGroup" | Should Be $true
+            $result -contains "Rename-LocalGroup" | Should Be $true
+            $result -contains "Remove-LocalGroup" | Should Be $true
+        }
     }
 
-    It "Creates New-LocalGroup with name and description" {
-        $result = New-LocalGroup -Name TestGroupAddRemove -Description "Test Group New 1 Description"
+    Describe "Validate simple New-LocalGroup" -Tags "CI", "Feature" {
```
Comment (vors) [inline test/powershell/Modules/Microsoft.PowerShell.LocalAccounts/Pester.Command.Cmdlets.LocalAccounts.LocalGroup.Tests.ps1]: Same comment

**Review by vors — CHANGES_REQUESTED** [review]
Please, remove duplicated priority tags

File: test/powershell/Modules/Microsoft.PowerShell.LocalAccounts/Pester.Command.Cmdlets.LocalAccounts.LocalGroup.Tests.ps1 (hunk 1 above)
Comment (mirichmo) [inline test/powershell/Modules/Microsoft.PowerShell.LocalAccounts/Pester.Command.Cmdlets.LocalAccounts.LocalGroup.Tests.ps1]: I got most of them and missed those two. They are fixed now.

**Review by vors — APPROVED** [review]

**Pull request merged**